src/Control/Effect/Cull.hs

```diff
@@ -51,18 +52,20 @@ cull m = send (Cull m pure)
 --
 --   prop> run (runNonDet (runCull (pure a <|> pure b))) === [a, b]
 runCull :: Alternative m => CullC m a -> m a
-runCull (CullC m) = runNonDetC (runReader False m) ((<|>) . pure) empty
+runCull (CullC m) = runNonDetC (runReader False m) (<|>) pure empty
 
 newtype CullC m a = CullC { runCullC :: ReaderC Bool (NonDetC m) a }
   deriving (Applicative, Functor, Monad, Fail.MonadFail, MonadFix, MonadIO)
 
 instance Alternative (CullC m) where
   empty = CullC empty
   {-# INLINE empty #-}
-  l <|> r = CullC $ ReaderC $ \ cull -> NonDetC $ \ cons nil -> do
-    runNonDetC (runReader cull (runCullC l))
-      (\ a as -> cons a (if cull then nil else as))
-      (runNonDetC (runReader cull (runCullC r)) cons nil)
+  l <|> r = CullC $ ReaderC $ \ cull ->
+    if cull then
+      NonDetC $ \ fork leaf nil ->
+        runNonDetC (runReader cull (runCullC l)) fork leaf (runNonDetC (runReader cull (runCullC r)) fork leaf nil)
+    else
+      runReader cull (runCullC l) <|> runReader cull (runCullC r)
```

Comment:
When we aren’t inside a `cull`, then we just alternate as normal for the underlying `NonDetC`.